From 3978222106a74b33f3efb8d07435f7165968625c Mon Sep 17 00:00:00 2001
From: Daniel Colascione <dancol@dancol.org>
Date: Sun, 23 Mar 2014 03:24:51 -0700
Subject: The compose key should now work better with UIM (1/4).

This upstream patch has been added:

  Work around flaky XIM modules

Origin: upstream, commit: r116856.1.1, 3ee4c51c34a049dc4bd36fb04c33cdf9a85e5ffa
Added-by: Rob Browning <rlb@defaultvalue.org>
Bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=753534
---
 src/ChangeLog |  5 +++++
 src/xfns.c    | 10 ++++++++++
 2 files changed, 15 insertions(+)

diff --git a/src/ChangeLog b/src/ChangeLog
index 01acb02..271d81d 100644
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,3 +1,8 @@
+2014-03-23  Daniel Colascione  <dancol@dancol.org>
+
+	* xfns.c (create_frame_xic): If XCreateIC fails, try again without
+	XNStatusAttributes; works around flaky XIM modules, apparently.
+
 2013-08-21  Paul Eggert  <eggert@cs.ucla.edu>
 
 	* process.c (allocate_pty) [PTY_OPEN]: Set fd's FD_CLOEXEC flag.
diff --git a/src/xfns.c b/src/xfns.c
index 656b71f..f1d4a48 100644
--- a/src/xfns.c
+++ b/src/xfns.c
@@ -2221,6 +2221,16 @@ create_frame_xic (struct frame *f)
 		       XNStatusAttributes, status_attr,
 		       XNPreeditAttributes, preedit_attr,
 		       NULL);
+
+      /* Some input methods don't support a status pixel.  */
+      if (xic == NULL)
+        xic = XCreateIC (xim,
+                         XNInputStyle, xic_style,
+                         XNClientWindow, FRAME_X_WINDOW (f),
+                         XNFocusWindow, FRAME_X_WINDOW (f),
+                         XNPreeditAttributes, preedit_attr,
+                         NULL);
+
       XFree (preedit_attr);
       XFree (status_attr);
     }
